File: R/asAnnulePaire.R
asAnnulePaire <- function(AS, k) {
  # AS$GS(v,v) est la transformation chol() (en triangle supérieur) de R(v,v)
  # k est un scalaire pointant dans AS$Cpaires(np,2)
  # Optimise les variables de la rangée k de AS$Cpaires pour en minimiser le signal
  # Remplit AS$Crit[k], AS$Ppaires[k], AS$Corr[,k], AS$Prob[k]
  # AS$absLogPo sert à indiquer une annulation pas réussie (poids très petit ou très grand)
  # AS$Crit[k] sera max(abs(corrélations du contraste))
  # AS$Prob pourra être transformé copr devenir des distances pour l'analyse en grappes
  cible <- AS$pertinent
  melange <- AS$Cpaires[, k]
  pd <- probNonDoublet(AS, melange)
  r <- AS$R[melange[1],melange[2]]  # corrélation des variables mêmes
  if (pd < 0.05) {  # si la paire a d'autres corrélations
    ap <- crit_paire_initiale(AS,melange[1],melange[2])
  } else {    # si la paire semble dépendre d'un facteur doublet
    AS$doublet <- rbind(AS$doublet, melange)  # documenter le doublet dans AS
    out <- prodCorr(r,AS$N)
    cont <- AS$GS[,melange] %*% c(1,-1)
    corr <- t(sc1(cont)) %*% sc1(AS$GS[,-melange])
    ap <- list(prob=out$p,po=sign(r),crit=max(abs(corr)),corr=corr)
    ap$po <- sign(r)  # Pour un doublet, les 2 variables auront le même poids en val. abs.
  }
  if (abs(log(abs(ap$po)))>5){
    AS$Prob[k] <- .1^abs(log(abs(ap$po)))
  }
  else
    AS$Prob[k] <- ap$prob
  AS$Crit[k] <- ap$crit
  AS$Corr[,k] <- ap$corr
  if (any(is.nan(ap$corr))) browser()
  AS$Ppaires[k] <- ap$po
  if (r*ap$po>0) {
    s1 <- sqrt(r*ap$po)
    s2 <- r/s1
  } else {
      s1 <- s2 <- NA
    }
  AS$satPaires[,k] <- c(s1,s2)
  return(AS)
}

# AS$Crit <- rep(0, nc)      # max(abs(Corr))
# AS$Prob <- rep(0, nc)      # corrigé pour le nombre de corrélations
# AS$Ppaires <- rep(0, nc)   # les poids d'annulation
# AS$Corr <- matrix(0, AS$nv-2, nc)
# AS$doublet <- NULL


# CritLog <- function(px, G, combine, cible, K, to.opt = TRUE) {
#   pp <- c(exp(px), K)
#   SP <- G[, combine] %*% pp
#   SP <- SP %*% solve(sqrt(t(SP) %*% (SP)))
#   corr <- t(SP) %*% G
#   corr[setdiff(1:length(corr), setdiff(cible, combine))] <- 0
#   crit <- max(corr^2)
#   if (is.complex(crit)) {
#     crit <- 99e9
#   }
#   if(to.opt){
#     out <- crit
#   } else {
#     out <- list(crit = crit, corr = corr)
#   }
#   return(out)
# }

File: Matlab/SCRoF_mars25/asAnnulePaire.m
function AS=asAnnulePaire(AS,k)
% AS=asAnnulePaire(AS,k);
% AS.GS(v,v) est la GrammSchmidt (en triangle supérieur) de R(v,v)
% k est un scalaire
% Optimise les variables de la rangée k de AS.Cpaire pour en
% minimiser le signal
% remplit AS.Crit(k), AS.Ppaires(k) et AS.Corr(:,k)
% AS.Crit(k) est déjà le X2: N-1 fois la somme des corrélations au carré

% limite=.98; % limite sur la plus grande saturation en valeur absolue
G=AS.GS;
cible=AS.pertinent;
melange=AS.Cpaires(k,:);
lP=0;
pd=probNonDoublet(AS,melange);
if pd<.05
    r=AS.R(melange(1),melange(2));
    K=-sign(r);  % le poids fixe pour la 2e variable
    lP=0;  % logarithme de P=1
    options=optimset('MaxFunEvals',1e4,'MaxIter',1e4,'TolFun',1e-4,'TolX',1e-4);
    lP=fminsearch(@(lP) CritLog(lP,G,melange,cible,K),lP,options);
    P=exp(lP);
    % r=AS.R(melange(1),melange(2));
    % ri=r/limite;
    % rs=limite/r;
    % if r>0
    %     if P>rs
    %         P=rs;
    %     elseif P<ri
    %         P=ri;
    %     end
    % else
    %     if P<rs
    %         P=rs;
    %     elseif P>ri
    %         P=ri;
    %     end
    % end
else
    AS.doublet=[AS.doublet;melange];
    K=1;
end
[~,corr]=CritLog(lP,G,melange,cible,K);
AS.Crit(k)=corr*corr'*(AS.N-1);
AS.Corr(:,k)=corr;
AS.Ppaires(k)=exp(lP);
end

function [crit,corr]=CritLog(px,G,combine,cible,K)
pp=[exp(px);K];
SP=G(:,combine)*pp;
SP=SP./sqrt(SP'*SP); 
corr=SP'*G;   % corr(1,v)
corr(setdiff(1:numel(corr),setdiff(cible,combine)))=0;
crit=max(corr.^2);
if ~isreal(crit)
    crit=99e9;
end
end